byron/ledger/executable-spec/test/Generator.hs

```diff
@@ -0,0 +1,109 @@
+module Generator where
+
+import           Data.Bits (shiftL)
+import           Data.Word (Word64)
+import           Hedgehog
+import qualified Hedgehog.Gen    as Gen
+import qualified Hedgehog.Range  as Range
+import           Ledger.Core (Slot(..), SlotCount(..), BlockCount(..))
+import           Ledger.GlobalParams (k)
+import           Ledger.Update (ProtVer(..), PParams(..))
+import           Numeric.Natural (Natural)
+
+-- | Generates a slot within the given bound
+slotGen :: Word64 -> Word64 -> Gen Slot
+slotGen lower upper = do
+  Slot <$> Gen.word64 (Range.linear lower upper)
+
+-- | Generates a protocol version
+protVerGen :: Gen ProtVer
+protVerGen = do
+  pvMin  <- Gen.integral (Range.linear (0 :: Natural) 100)
+  pvDiff <- Gen.integral (Range.linear (0 :: Natural) 20)
+  let pvMax = pvMin + pvDiff
+  pvAlt  <- Gen.integral (Range.linear (0 :: Natural) 100)
+  return $ ProtVer pvMin pvMax pvAlt
+
+-- | Generates a Double in the inclusive range [0, 1]
+doubleGen :: Gen Double
+doubleGen = do
+  b <- Gen.word64 (Range.linear 1 hi)
+  a <- Gen.word64 (Range.linear 0 b )
+  return $ (fromIntegral a) / (fromIntegral b)
+ where
+  hi = 2 `shiftL` 48 :: Word64
+
+-- | Generates valid protocol parameters
+pparamsGen :: Gen PParams
+pparamsGen = do
+  maxBkSz         <- Gen.integral (Range.linear 1 1000)
+  maxHdrSz        <- Gen.integral (Range.linear 1 maxBkSz)
+  maxTxSz         <- Gen.integral (Range.linear 1 maxBkSz)
+  maxPropSz       <- Gen.integral (Range.linear 1 maxBkSz)
+  bkSgnCntT       <- doubleGen
+  bkSlotsPerEpoch <- SlotCount <$> Gen.word64 (Range.linear 1 21600)
+  upTtl           <- SlotCount <$> Gen.word64
+    (Range.linear 1 (unSlotCount bkSlotsPerEpoch))
+  scriptVersion   <- Gen.integral (Range.linear (0 :: Natural) 1000)
+  cfmThd          <- Gen.integral (Range.linear 0 1000)
+  upAdptThd       <- Gen.integral (Range.linear 1 100)
+  stableAfter     <- BlockCount <$> Gen.word64
+    (Range.linear 1 (unSlotCount bkSlotsPerEpoch))
+  return $
+    PParams
+      maxBkSz
+      maxHdrSz
+      maxTxSz
+      maxPropSz
+      bkSgnCntT
+      bkSlotsPerEpoch
+      upTtl
+      scriptVersion
+      cfmThd
+      upAdptThd
+      stableAfter
+
+data PVBUMPEnv = PVBUMPEnv
```

Comment:
This should be type `PVBUMPEnv = Environment PVBUMP` (or just use `Environment PVBUMP`).

Reply:
Actually I'm thinking of changing the type in the STS itself. The reason is that wrappers like these help in both communicating to the developer what this is about (compared to a pair or a product in general) and also helps to avoid bugs due to otherwise hard to distinguish types.

Reply:
>  I'm thinking of changing the type in the STS itself

I'd suggest not to do that in this PR. Also I don't know what do you plan on changing the type to. For the chain executable spec I thought about using tuples, since this results in the formal and executable spec having a very similar appearance. However compiler errors are harder to debug.

Whatever you do, please include @nc6 and me in the initial design.

Reply:
Sure, I'll leave it that way in `tests` only.